src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -599,7 +602,20 @@ Pipes MergeTreeDataSelectExecutor::readFromParts(
                         index_and_condition.first, index_and_condition.second, part, ranges.ranges, settings, reader_settings, log);
 
             if (!ranges.ranges.empty())
+            {
+                if (settings.read_overflow_mode == OverflowMode::THROW && settings.max_rows_to_read)
+                {
+                    /// Fail fast if estimated number of rows to read exceeds the limit
+                    size_t total_rows_estimate = total_rows.fetch_add(ranges.getRowsCount());
+                    if (total_rows_estimate > settings.max_rows_to_read)
+                        throw Exception(
+                            "Limit for rows exceeded, max rows: " + formatReadableQuantity(settings.max_rows_to_read)
+                            + ", current rows: " + formatReadableQuantity(total_rows_estimate),
```

Comment:
nit: total_row_estimate in this error message might be a bit miss-leading as it can be interpreted by the user as "limit 100 but need to read $total_row_estimate". Is it much more expensive if this is moved after all process_part are executed so that the actual number of estimated rows could be printed in the error (this would be consistent with current limit logic irc)? If it would be expensive I'd just remove this part of the message.

Reply:
`current rows` -> `estimated rows to read (at least)`
And also would be nice to add the name of the setting:
`Limit for rows (controlled by `max_rows_to_read` setting) exceeded`